crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2490,130 +2447,371 @@ impl<'db> Type<'db> {
+                        Signature::new(
+                            Parameters::new([Parameter::new(
+                                Some(Name::new_static("o")),
+                                Some(Type::any()),
+                                ParameterKind::PositionalOnly { default_ty: None },
+                            )]),
+                            Some(KnownClass::Type.to_instance(db)),
+                        ),
+                        Signature::new(
+                            Parameters::new([
+                                Parameter::new(
+                                    Some(Name::new_static("o")),
+                                    Some(Type::any()),
+                                    ParameterKind::PositionalOnly { default_ty: None },
+                                ),
+                                Parameter::new(
+                                    Some(Name::new_static("bases")),
+                                    Some(Type::any()),
+                                    ParameterKind::PositionalOnly { default_ty: None },
+                                ),
+                                Parameter::new(
+                                    Some(Name::new_static("dict")),
+                                    Some(Type::any()),
+                                    ParameterKind::PositionalOnly { default_ty: None },
+                                ),
+                            ]),
+                            Some(KnownClass::Type.to_instance(db)),
+                        ),
+                    ],
+                );
+                Signatures::single(signature)
+            }
+
+            // TODO annotated return type on `__new__` or metaclass `__call__`
+            // TODO check call vs signatures of `__new__` and/or `__init__`
+            Type::ClassLiteral(ClassLiteralType { .. }) => {
+                let signature = CallableSignature::single(
+                    callable_ty,
+                    self,
+                    Signature::new(Parameters::gradual_form(), self.to_instance(db)),
+                );
+                Signatures::single(signature)
+            }
+
+            Type::SubclassOf(subclass_of_type) => match subclass_of_type.subclass_of() {
+                ClassBase::Dynamic(dynamic_type) => Type::Dynamic(dynamic_type)
+                    .signatures(db, callable_ty)
+                    .clone(),
+                ClassBase::Class(class) => Type::class_literal(class)
+                    .signatures(db, callable_ty)
+                    .clone(),
+            },
+
+            Type::Instance(_) => {
+                // Note that for objects that are callable via a `__call__` method, we will get the
+                // signature of the dunder method, but will pass in the type of the object as the
+                // "callable type". That ensures that we get errors like "`X` is not callable"
+                // instead of "`<type of illegal '__call__'>` is not callable".
```

Comment:
A small suggested tweak to the comment here: if a type has `__call__` set to `None`, it is not callable from the language's perspective: in order for an object to be callable, it has to have a `__call__` method that is itself callable. So it's a _little_ confusing to say here that an object might be "callable via a `__call__` method" but then later in the same paragraph say that attempting to call the `__call__` method might then cause us to emit an error saying that "`__call__` is not callable".

```suggestion
                // Note that for objects that have a (possibly not callable!) `__call__` attribute,
                // we will get the signature of the `__call__` attribute, but will pass in the type
                // of the original object as the "callable type". That ensures that we get errors
                // like "`X` is not callable" instead of "`<type of illegal '__call__'>` is not
                // callable".
```

Reply:
Thanks!